### ISSUE TITLE
feed sort order with ttrss sync

### ISSUE THREAD

**Issue by serpent7776** [newsboat/newsboat#161]
In recent newsboat version default feed sort order has been changed. I've set up synchronization with ttrss and I set my preferred feed order on server side. now newsboat is changing my preferred sort order and this cannot be changed. Setting sort order to none does not help. I'd like to have an option to take feed sort order from ttrss server, as it was before.

**Newsboat version (copy from `newsboat -v`)**:
newsboat 2.11.1
System: FreeBSD 11.1-RELEASE-p9 (amd64)
Compiler: g++ 4.2.1 Compatible FreeBSD Clang 4.0.0 (tags/RELEASE_400/final 297347)
ncurses: ncurses 5.9.20140222 (compiled with 5.9)
libcurl: libcurl/7.59.0 OpenSSL/1.0.2o zlib/1.2.11 libidn2/2.0.4 nghttp2/1.31.0 (compiled with 7.59.0)
SQLite: 3.22.0 (compiled with 3.22.0)
libxml2: compiled with 2.9.7

Steps to reproduce the issue:

1. Set up ttrss support
2. launch newsboat
3. Feeds will be sorted by name

**Comment by Minoru** [comment]
Thanks for the issue. Yet I don't see any such changes in TT-RSS code. Maybe a move to nlohmann/json is the cause? @noctux, any comment on that?

**Comment by serpent7776** [comment]
I think that was a newsboat change, not tt-rss. 
That might be casued by this change in newsboat `Sort the file list in filebrowser (Nikos Tsipinakis)`

**Comment by Minoru** [comment]
Oh, sorry—I meant *our* code that deals with TT-RSS.

The change you mention is unrelated to the feedlist. I just took a look at the patch, and I can't conceive a way in which it could affect any other dialog.

**Comment by noctux** [comment]
I don't think that the json library is to blame. Instead, I believe that my recent modifications to the feed discovery are to blame for that. Up to the last release, newsboat retrieved the feeds by iterating over all categories and inserting the feeds in the order of the those categories and the order of the feeds inside those categories, as they are returned by the `getCategories` and `getFeeds` categories. Since the last release, we skip `getCategories` altogether and use `getFeeds` with the metacategory `-3 All feeds, excluding virtual feeds (e.g. Labels and such)` for all TTRSS-Instances with API-level 2 or above, which might result in a different ordering of the feeds w.r.t. the old variant. However, as far as I can see, the [official TTRSS-api documentation](https://git.tt-rss.org/git/tt-rss/wiki/ApiReference) does not mention that any kind of ordering is guaranteed for either of those two variants -- if it worked and actually resembled the actual server-side ordering, we would be relying on undocumented behaviour there.

But before we continue that branch of the discussion, lets first properly debug this:

@serpent7776: Could you try the following patch? If I'm right, then this patch should restore the old behaviour for you and confirm that the new feed-enumeration code is to blame for that:
[ttrss_debug.patch.txt](https://github.com/newsboat/newsboat/files/1909557/ttrss_debug.patch.txt)

**Comment by serpent7776** [comment]
@noctux your patch does indeed revert previous behaviour. 
So what would be the best option to keep tt-rss ordering in newsboat? Is it even possible?
I'm mostly concerned about ordering of directories. Ordering of feeds in any specific directory is not that much important for me.

**Comment by noctux** [comment]
On Sat, Apr 14, 2018 at 05:41:00PM +0000, serpent7776 wrote:
> @noctux your patch does indeed revert previous behaviour.

Ok, nice. Thanks for testing the patch! This means that the cause of the
problem is indeed the new feed-discovery code we've introduced in the
last release, as outlined above.

> So what would be the best option to keep tt-rss ordering in newsboat?
> Is it even possible?

Well, this depends. Given that the patch works, we could always add a
`ttrss-use-legacy-feed-enumeration` (or whatever we choose to call it)
configuration option, which restores the old behaviour with minimal
code increase, at least as long as we support TTRSS api-levels < 2.

However, as I've said, we are using undocumented behaviour, even in the
old case, because we require the server to hand out the list of
categories or feeds in a specific order (By the way, for my instance,
the orders do not really add up...). I don't really like relying on
that sort of thing, as such solutions tend to break one's back in the
long run.

However, there is a field called `order_id` in the return values of both
`getCategories` as well as `getFeeds`. It is not really documented
either, so I might be wrong, but the name and it's mapping to values for
my instance seems to suggest that it encodes the information on
directory order from the server side. (As a side note for whoever will
touch the code: As it's undocumented, we want to be defensive when using
it, and gracefully degrade if `order_id` is missing. Furthermore, it is
lacking for "special" categories like "-1/Special" or
"0/Uncategorized").
If my perception of the fields values was right, it should be possible
to use this field to order the feedlist for all api-levels in a clean
fashion - this is the solution I'd prefer, even though the code required
is more elaborate than the legacy technique outlined above.

I'll try to implement that solution so we can test if this actually
works out, but I can't give too strict guarantees on the timeframe. But
in the meantime, the patch posted above should suffice.
